### ISSUE TITLE
Getters run even when a field is not selected

### ISSUE THREAD

**Issue by Faibk** [Automattic/mongoose#5873]
When a field is deselected in a query its getter still runs . When using nested schemas this leads to an error due to the getters input being undefined. It is possible to check for undefined in the getter method, but this seems more like a hack to me. 
If the field is deselected and its value undefined, then the getter should not run as well.

Here is a small test case:

```javascript
it('Getter does only run when field is selected', function(done) {

        var getterRan = false;

        var testSchema = new Schema({
            name: String,
            skillflow: { type: [String], 'get': unstackScopes, 'set': stackScopes }
        }, {
            toObject: { getters: true },
            toJSON: { getters: true }
        });

        function stackScopes(skillflow) {
            console.log('setter run')
            return skillflow.map(scope => (scope.toString()));
        }

        function unstackScopes(skillflow) {
          console.log('getter run')
          getterRan = true;
            return skillflow.map(scope => scope.toString());
        }


        var Test = db.model('Test', testSchema);

        // Create a new batch of events with different kinds
        var program = {
            type: "Program",
            name: "Level1",
            skillflow: [
                "Test",
                "test"
            ]
        };

        Test.create(program).
        then(function(doc) {
            
            Test.findById(doc.id).select('-skillflow').exec(function(err,program){
              assert.equal(getterRan, false);
              done();
            })
        }).
        catch(done);
    });
```

**Comment by vkarpov15** [comment]
The getter isn't running when you do `findById().select()`, its running when you do `Test.create()`, and there's no projection there. The below script works fine.

```javascript
const assert = require('assert');
const mongoose = require('mongoose');

const Schema = mongoose.Schema;

var getterRan = false;
var testSchema = new Schema({
    name: String, 
    skillflow: { type: [String], 'get': unstackScopes, 'set': stackScopes }
}, {    
    toObject: { getters: true }, 
    toJSON: { getters: true }
});     

function stackScopes(skillflow) {
    console.log('setter run')
    return skillflow.map(scope => (scope.toString()));
}       

function unstackScopes(skillflow) {
  console.log('getter run')
  getterRan = true; 
    return skillflow.map(scope => scope.toString());
}       

mongoose.connect('mongodb://localhost:27017/test', { useMongoClient: true });
var Test = mongoose.model('Test', testSchema);

// Create a new batch of events with different kinds
var program = {
    type: "Program",
    name: "Level1",
    skillflow: [
        "Test", 
        "test"  
    ]       
};      

Test.create(program).
then(function(doc) {
    getterRan = false;
    console.log('-----------');   
    Test.findById(doc.id).select('-skillflow').exec(function(err,program){
      assert.equal(getterRan, false);
      console.log('done');
    })
}).
catch(error => console.error(error.stack));
```

Can you modify the above script to reproduce your issue?

**Comment by Faibk** [comment]
Thank You, seems like I did some overfitting here. I experienced the issue in a much more complex case and tried to boil it down to a test case. With your adaptation I cannot reproduce either. 
In my complex issue I have a nested schema with discriminators. When I exclude the skillflow field, I get an `undefined` error at the getter for this field, but in my understanding it should never run in the first place. 

Unfortunately I do not have time to build a more complex test case at the moment, so I am using the workaround from above and just return on undefined. Maybe I come back to this someday. I close this issue, as it's invalid for the moment.
